### PR TITLE
AttributedSpans equality check disregards order of attributions

### PR DESCRIPTION
**Pull request by britannio**
Closes https://github.com/superlistapp/super_editor/issues/270

### CODE REVIEW

**Comment by britannio** [comment]
Does VSCode have a setting to format lines at 120 characters?

**Comment by creativecreatorormaybenot** [comment]
@britannio it has nothing to do with VSCode. But you can change the `dart format` line wrap in settings :)

**Comment by britannio** [comment]
Cool, I created https://github.com/superlistapp/super_editor/pull/272 so I don't have to run the `dart format` command manually.

**Comment by creativecreatorormaybenot** [comment]
@britannio Of course not what I was suggesting :) Nice PR - I meant the user settings.

**Comment by matthew-carroll** [comment]
Can you add a test to the appropriate test file that verifies unordered equality?

**Comment by britannio** [comment]
@matthew-carroll Just added the tests!

File: super_editor/test/src/infrastructure/attributed_spans_test.dart [hunk 1]
```diff
@@ -477,6 +477,21 @@ void main() {
         expect(collapsedSpans[2].attributions.first, italics);
       });
     });
+
+    group('equality', () {
+      test('it determines that spans with equivalent content in a different order are equal', () {
```
Comment (matthew-carroll) [inline super_editor/test/src/infrastructure/attributed_spans_test.dart:482]: The wording in this description is a bit passive. How about:
it is equal to another AttributedSpans with equivalent markers that are stored in a different order

File: super_editor/test/src/infrastructure/attributed_spans_test.dart [hunk 2]
```diff
@@ -477,6 +477,21 @@ void main() {
         expect(collapsedSpans[2].attributions.first, italics);
       });
     });
+
+    group('equality', () {
```
Comment (matthew-carroll) [inline super_editor/test/src/infrastructure/attributed_spans_test.dart:481]: If we're going to write a test for different marker order, then we should probably write a test for the standard case where the marker order is the same.

Also, how about a test for empty `AttributedSpans`s?

And, when it comes to proving things, the converse is important, too. Consider writing one or more tests that demonstrate non-equality, to make sure that equality doesn't always return `true`.

**Review by matthew-carroll — CHANGES_REQUESTED** [review]
This is a good start on tests, but we should be more holistic with the testing range if we're going to take the time to add them.

**Comment by britannio** [comment]
Thanks for the feedback! Hopefully, I've addressed it all.

File: super_editor/test/src/infrastructure/attributed_spans_test.dart [hunk 3]
```diff
@@ -477,6 +478,42 @@ void main() {
...
+      const boldStart = SpanMarker(attribution: bold, offset: 0, markerType: SpanMarkerType.start);
```
Comment (matthew-carroll) [inline super_editor/test/src/infrastructure/attributed_spans_test.dart:483]: I'll accept this test as-is so that we can merge it, but in general I think that test configurations should go within the tests. Right now it doesn't look too bad to have these 4 properties here, but imagine that this group of tests grows to be 50 tests. How many different configurations might we have in that case? They'd all accumulate here and become a mess. So, for tests, I recommend re-stating the conditions in each test. In rare cases, you might move some of the setup generation into a method that a test invokes, but even that should be rare.

**Review by matthew-carroll — APPROVED** [review]
LGTM. Thanks for the implementation and for following up with tests.
